File: Sources/RakaatCore/PoseFeatures.swift
import Foundation

/// Scale-invariant features derived from one frame's joints. All vertical
/// distances are normalized by torso length (shoulder→hip), which stays roughly
/// constant across postures, so thresholds don't depend on the person's distance
/// from the camera.
public struct PoseFeatures: Equatable, Sendable {
    /// Angle of the spine (hip→shoulder vector) from vertical, in degrees 0...180.
    /// ~0 = upright (standing/sitting), ~90 = horizontal (ruku/sujud).
    public let torsoAngle: Double
    /// (head.y − ankle.y) / torsoLength. How high the head sits above the floor.
    /// High when standing/ruku, low in sujud.
    public let headAboveAnkle: Double
    /// (hip.y − ankle.y) / torsoLength. High when standing, low when sitting.
    public let hipAboveAnkle: Double
    /// (head.y − hip.y) / torsoLength. Positive standing, ~0 ruku, negative sujud.
    public let headAboveHip: Double

    public init(torsoAngle: Double, headAboveAnkle: Double, hipAboveAnkle: Double, headAboveHip: Double) {
        self.torsoAngle = torsoAngle
        self.headAboveAnkle = headAboveAnkle
        self.hipAboveAnkle = hipAboveAnkle
        self.headAboveHip = headAboveHip
    }

    /// Compute features from joints. Returns nil if required joints are missing
    /// or the body has no measurable scale.
    public static func from(_ j: BodyJoints) -> PoseFeatures? {
        guard let shoulder = j.shoulderMid,
              let hip = j.hipMid,
              let ankle = j.ankleMid,
              let head = j.head else { return nil }

        let torso = shoulder - hip
        let torsoLength = torso.length
        guard torsoLength > 1e-6 else { return nil }

        // Angle from vertical (0,1): cos = torso.y / |torso|.
        let cosA = max(-1.0, min(1.0, torso.y / torsoLength))
        let angle = acos(cosA) * 180.0 / Double.pi

        return PoseFeatures(
            torsoAngle: angle,
            headAboveAnkle: (head.y - ankle.y) / torsoLength,
            hipAboveAnkle: (hip.y - ankle.y) / torsoLength,
            headAboveHip: (head.y - hip.y) / torsoLength
        )
    }
}

File: web/src/core/poseFeatures.ts
import {
  BodyJoints,
  Point2D,
  ankleMid,
  head,
  hipMid,
  length,
  shoulderMid,
  sub,
} from "./geometry.js";

/** Scale-invariant features derived from one frame's joints. Vertical distances
 *  are normalized by torso length (shoulder→hip), which stays roughly constant
 *  across postures, so thresholds don't depend on camera distance. */
export interface PoseFeatures {
  /** Spine angle from vertical, degrees 0..180. ~0 upright, ~90 horizontal. */
  torsoAngle: number;
  /** (head.y − ankle.y)/torsoLength. High standing/ruku, low in sujud. */
  headAboveAnkle: number;
  /** (hip.y − ankle.y)/torsoLength. High standing, low sitting. */
  hipAboveAnkle: number;
  /** (head.y − hip.y)/torsoLength. Positive standing, ~0 ruku, negative sujud. */
  headAboveHip: number;
}

/** Returns null if required joints are missing or the body has no scale. */
export function poseFeaturesFrom(j: BodyJoints): PoseFeatures | null {
  const shoulder = shoulderMid(j);
  const hip = hipMid(j);
  const ankle = ankleMid(j);
  const h = head(j);
  if (!shoulder || !hip || !ankle || !h) return null;

  const torso: Point2D = sub(shoulder, hip);
  const torsoLength = length(torso);
  if (torsoLength <= 1e-6) return null;

  // Angle from vertical (0,1): cos = torso.y / |torso|.
  const cosA = Math.max(-1, Math.min(1, torso.y / torsoLength));
  const torsoAngle = (Math.acos(cosA) * 180) / Math.PI;

  return {
    torsoAngle,
    headAboveAnkle: (h.y - ankle.y) / torsoLength,
    hipAboveAnkle: (hip.y - ankle.y) / torsoLength,
    headAboveHip: (h.y - hip.y) / torsoLength,
  };
}
